File: Python/24_longest_common_subsequence.py
import sys


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def longest_common_subsequence(s1, s2, n):
    longest_path = [[0] * (n + 1) for _ in range(n+1)]
    backtrack =  [[0] * n for _ in range(n)]
    for ix in range(n):
        for jx in range(n):
            longest_path[ix+1][jx+1] = max(longest_path[ix][jx+1],
                                           longest_path[ix+1][jx],
                                           longest_path[ix][jx] + int(s1[ix] == s2[jx]))
            if longest_path[ix+1][jx+1] == longest_path[ix][jx+1]:
                backtrack[ix][jx] = -1
            elif longest_path[ix+1][jx+1] == longest_path[ix+1][jx]:
                backtrack[ix][jx] = 1
            elif (longest_path[ix+1][jx+1] == longest_path[ix][jx] + 1) and s1[ix] == s2[jx]:
                backtrack[ix][jx] = 0
    return output_longest_common_subsequence(backtrack, n)


def output_longest_common_subsequence(backtrack, n):
    ixs1 = []
    ixs2 = []
    ix, jy = n, n
    while ix > 0 and jy > 0:
        if backtrack[ix-1][jy-1] == 0:
            ixs1.append(ix-1)
            ixs2.append(jy-1)
            ix -= 1
            jy -= 1
        elif backtrack[ix-1][jy-1] == -1:
            ix -= 1
        else:
            jy -= 1
    return ixs1[::-1], ixs2[::-1]


def stringify_array(array):
    return ' '.join([str(item) for item in array])


def main():
    reader = sys.stdin
    n = int(next(reader))
    s1 = convert_to_intlist(next(reader))
    s2 = convert_to_intlist(next(reader))
    ixs1, ixs2 = longest_common_subsequence(s1, s2, n)
    print(len(ixs1))
    print(stringify_array(ixs1))
    print(stringify_array(ixs2))


if __name__ == "__main__":
    main()

File: Haskell/24-longest-common-subsequence.hs
import Control.Monad (replicateM)
import Data.Array (Array, listArray, range, (!))


convertToIntList :: String -> [Int]
convertToIntList = map read . words


backtrackLCS :: Array (Int, Int) Int -> Int -> ([Int], [Int])
backtrackLCS backtrack n = go [] [] n n
    where
        go :: [Int] -> [Int] -> Int -> Int -> ([Int], [Int])
        go ixs1 ixs2 i j
            | i <= 0 || j <= 0 = (ixs1, ixs2)
            | backtrack ! (i, j) == 0 = go ((i - 1) : ixs1) ((j - 1) : ixs2) (i - 1) (j - 1)
            | backtrack ! (i, j) == -1 = go ixs1 ixs2 (i - 1) j
            | otherwise = go ixs1 ixs2 i (j - 1)


longestCommonSubsequence :: Array Int Int -> Array Int Int -> Int -> ([Int], [Int])
longestCommonSubsequence s1 s2 n =
    let fillTables :: Int -> Int -> (Int, Int)
        fillTables 0 _ = (0, 0)
        fillTables _ 0 = (0, 0)
        fillTables i j =
            let mismatch = if s1 ! (i-1) == s2 ! (j-1) then 1 else 0
                path = maximum [longestPath ! (i-1, j), longestPath ! (i, j-1), longestPath ! (i-1, j-1) + mismatch]
                direction
                    | path == longestPath ! (i-1, j) = -1
                    | path == longestPath ! (i, j-1) = 1
                    | otherwise = 0
            in (path, direction)
        bounds = ((0, 0), (n, n))
        (longestPath, backtrack) =
            let (lp, bt) = unzip $ [fillTables i j | (i, j) <- range bounds]
            in (listArray bounds lp, listArray bounds bt)
    in backtrackLCS backtrack n


main :: IO()
main = do
    n <- readLn
    s1 <- fmap (listArray (0, n - 1) . convertToIntList) getLine
    s2 <- fmap (listArray (0, n - 1) . convertToIntList) getLine
    let (ixs1, ixs2) = longestCommonSubsequence s1 s2 n
    print $ length ixs1
    putStrLn $ unwords $ map show ixs1
    putStrLn $ unwords $ map show ixs2
